expression/src/main/java/io/smallrye/common/expression/Expression.java

```diff
@@ -485,15 +507,30 @@ private static Node parseString(Itr itr, final boolean allowExpr, final boolean
                             }
                             //throw Assert.unreachableCode();
                         }
+                        case '\\': {
+                            if (flags.contains(Flag.NO_$$)) {
+                                if (itr.hasNext() && itr.peekNext() == '$') {
+                                    list.add(new LiteralNode(itr.getStr(), start, idx));
+                                    start = itr.getNextIdx();
+                                    itr.next();
+                                    continue;
+                                }
+                            }
+                        }
                         default: {
+                            // enqueue what we have acquired so far
+                            if (idx > start) {
+                                // TP 4
+                                list.add(new LiteralNode(itr.getStr(), start, idx - 1));
+                            }
                             // $ followed by anything else
                             if (flags.contains(Flag.MINI_EXPRS)) {
                                 // TP 25
                                 list.add(new ExpressionNode(false, new LiteralNode(itr.getStr(), idx, itr.getNextIdx()),
                                         Node.NULL));
                                 start = itr.getNextIdx();
                                 continue;
-                            } else if (flags.contains(Flag.LENIENT_SYNTAX)) {
+                            } else if (flags.contains(Flag.LENIENT_SYNTAX) || flags.contains(Flag.NO_$$)) {
```

Comment:
Why do we care about `NO_$$` here?

Reply:
In the case of expressions like `$$foo`, when we reevalute the second '$' we fall in the `// $ followed by anything else` case.

Reply:
OK, I guess this was a perf optimization. It might be cleaner if for this case we enqueue what we have so far (before the `$`), add a literal `$` item, and then continue from initial, which would flatten a couple of other cases too. (This might be similar to what you had before; if so, sorry about that... 🙁)

Reply:
No worries, all good :)